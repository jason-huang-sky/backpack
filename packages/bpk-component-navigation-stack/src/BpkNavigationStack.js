/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import { cssModules } from 'bpk-react-utils';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import React, { Component, cloneElement, Children, type Element } from 'react';

import STYLES from './BpkNavigationStack.scss';

const getClassName = cssModules(STYLES);

type Props = {
  initialViews: Array<Element<any>>,
  className: ?string,
};

type State = {
  views: Array<Element<any>>,
};

class BpkNavigationStack extends Component<Props, State> {
  static defaultProps = {
    className: null,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      views: this.props.initialViews,
    };
  }

  pushView(view: Element<any>) {
    this.setState(prevState => {
      const views = prevState.views.slice();
      views.push(view);

      return {
        views,
      };
    });
  }

  popView() {
    this.setState(prevState => {
      const views = prevState.views.slice();
      views.pop();

      return {
        views,
      };
    });
  }

  render() {
    const { initialViews, className, ...rest } = this.props;
    const classNames = [getClassName('bpk-navigation-stack')];

    if (className) {
      classNames.push(className);
    }

    const viewClassNames = {
      enter: getClassName('bpk-navigation-stack__view--enter'),
      enterActive: getClassName('bpk-navigation-stack__view--enter-active'),
      enterDone: getClassName('bpk-navigation-stack__view--enter-done'),
      exit: getClassName('bpk-navigation-stack__view--exit'),
      exitActive: getClassName('bpk-navigation-stack__view--exit-active'),
      exitDone: getClassName('bpk-navigation-stack__view--exit-done'),
    };

    return (
      <TransitionGroup className={classNames.join(' ')} {...rest}>
        {Children.map(this.state.views, view => (
          <CSSTransition timeout={500} classNames={viewClassNames}>
            <div className={getClassName('bpk-navigation-stack__view')}>
              {cloneElement(view, { navigationController: this })}
            </div>
          </CSSTransition>
        ))}
      </TransitionGroup>
    );
  }
}

export default BpkNavigationStack;
