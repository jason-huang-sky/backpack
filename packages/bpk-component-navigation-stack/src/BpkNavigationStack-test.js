/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import BpkNavigationStack from './BpkNavigationStack';

describe('BpkNavigationStack', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<BpkNavigationStack views={[<div />]} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should set "aria-hidden" to true on non active views', () => {
    const tree = renderer
      .create(<BpkNavigationStack views={[<div />, <div />, <div />]} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should set "tabbable" to true on the last view', () => {
    const tree = shallow(
      <BpkNavigationStack views={[<span />, <span />, <span />]} />,
    );
    const containers = tree.find('TabbableContainer');
    expect(containers.at(0).props().tabbable).toBe(false);
    expect(containers.at(1).props().tabbable).toBe(false);
    expect(containers.at(2).props().tabbable).toBe(true);
  });

  it('should set "autoFocus" to true on the last view, after the first render', () => {
    const tree = shallow(
      <BpkNavigationStack views={[<span />, <span />]} autoFocusNextView />,
    );

    let containers = tree.find('TabbableContainer');
    expect(containers.at(0).props().autoFocus).toBe(false);
    expect(containers.at(1).props().autoFocus).toBe(false);

    tree.setProps({ views: [<span />, <span />, <span />] });

    containers = tree.find('TabbableContainer');
    expect(containers.at(0).props().autoFocus).toBe(false);
    expect(containers.at(1).props().autoFocus).toBe(false);
    expect(containers.at(2).props().autoFocus).toBe(true);
  });
});
