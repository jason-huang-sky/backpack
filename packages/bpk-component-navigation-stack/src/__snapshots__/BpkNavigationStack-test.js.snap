// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkNavigationStack should render correctly 1`] = `
<div
  className="bpk-navigation-stack"
>
  <div
    className="bpk-navigation-stack__view"
  >
    <div
      navigationController={
        BpkNavigationStack {
          "_reactInternalInstance": ReactCompositeComponentWrapper {
            "_calledComponentWillUnmount": false,
            "_compositeType": 0,
            "_context": Object {},
            "_currentElement": <BpkNavigationStack
              className={null}
              initialViews={
                Array [
                  <div />,
                ]
              }
            />,
            "_debugID": 1,
            "_hostContainerInfo": Object {
              "createNodeMock": [Function],
            },
            "_hostParent": null,
            "_instance": [Circular],
            "_mountImage": null,
            "_mountIndex": 0,
            "_mountOrder": 2,
            "_pendingCallbacks": null,
            "_pendingElement": null,
            "_pendingForceUpdate": false,
            "_pendingReplaceState": false,
            "_pendingStateQueue": null,
            "_renderedComponent": ReactCompositeComponentWrapper {
              "_calledComponentWillUnmount": false,
              "_compositeType": 0,
              "_context": Object {},
              "_currentElement": <TransitionGroup
                childFactory={[Function]}
                className="bpk-navigation-stack"
                component="div"
              >
                <CSSTransition
                  classNames={
                    Object {
                      "enter": "bpk-navigation-stack__view--enter",
                      "enterActive": "bpk-navigation-stack__view--enter-active",
                      "enterDone": "bpk-navigation-stack__view--enter-done",
                      "exit": "bpk-navigation-stack__view--exit",
                      "exitActive": "bpk-navigation-stack__view--exit-active",
                      "exitDone": "bpk-navigation-stack__view--exit-done",
                    }
                  }
                  timeout={500}
                >
                  <div
                    className="bpk-navigation-stack__view"
                  >
                    <div
                      navigationController={[Circular]}
                    />
                  </div>
                </CSSTransition>
              </TransitionGroup>,
              "_debugID": 2,
              "_hostContainerInfo": Object {
                "createNodeMock": [Function],
              },
              "_hostParent": null,
              "_instance": TransitionGroup {
                "_reactInternalInstance": [Circular],
                "appeared": true,
                "context": Object {},
                "props": Object {
                  "childFactory": [Function],
                  "children": Array [
                    <CSSTransition
                      classNames={
                        Object {
                          "enter": "bpk-navigation-stack__view--enter",
                          "enterActive": "bpk-navigation-stack__view--enter-active",
                          "enterDone": "bpk-navigation-stack__view--enter-done",
                          "exit": "bpk-navigation-stack__view--exit",
                          "exitActive": "bpk-navigation-stack__view--exit-active",
                          "exitDone": "bpk-navigation-stack__view--exit-done",
                        }
                      }
                      timeout={500}
                    >
                      <div
                        className="bpk-navigation-stack__view"
                      >
                        <div
                          navigationController={[Circular]}
                        />
                      </div>
                    </CSSTransition>,
                  ],
                  "className": "bpk-navigation-stack",
                  "component": "div",
                },
                "refs": Object {},
                "state": Object {
                  "children": Object {
                    ".$.0": <CSSTransition
                      appear={undefined}
                      classNames={
                        Object {
                          "enter": "bpk-navigation-stack__view--enter",
                          "enterActive": "bpk-navigation-stack__view--enter-active",
                          "enterDone": "bpk-navigation-stack__view--enter-done",
                          "exit": "bpk-navigation-stack__view--exit",
                          "exitActive": "bpk-navigation-stack__view--exit-active",
                          "exitDone": "bpk-navigation-stack__view--exit-done",
                        }
                      }
                      enter={undefined}
                      exit={undefined}
                      in={true}
                      onExited={[Function]}
                      timeout={500}
                    >
                      <div
                        className="bpk-navigation-stack__view"
                      >
                        <div
                          navigationController={[Circular]}
                        />
                      </div>
                    </CSSTransition>,
                  },
                },
                "updater": Object {
                  "enqueueCallback": [Function],
                  "enqueueCallbackInternal": [Function],
                  "enqueueElementInternal": [Function],
                  "enqueueForceUpdate": [Function],
                  "enqueueReplaceState": [Function],
                  "enqueueSetState": [Function],
                  "isMounted": [Function],
                  "validateCallback": [Function],
                },
              },
              "_mountImage": null,
              "_mountIndex": 0,
              "_mountOrder": 3,
              "_pendingCallbacks": null,
              "_pendingElement": null,
              "_pendingForceUpdate": false,
              "_pendingReplaceState": false,
              "_pendingStateQueue": null,
              "_renderedComponent": <div
                className="bpk-navigation-stack"
              >
                <div
                  className="bpk-navigation-stack__view"
                >
                  <div
                    navigationController={[Circular]}
                  />
                </div>
              </div>,
              "_renderedNodeType": 0,
              "_rootNodeID": 0,
              "_topLevelWrapper": null,
              "_updateBatchNumber": null,
              "_warnedAboutRefsInRender": false,
            },
            "_renderedNodeType": 1,
            "_rootNodeID": 0,
            "_topLevelWrapper": ReactCompositeComponentWrapper {
              "_calledComponentWillUnmount": false,
              "_compositeType": 0,
              "_context": Object {},
              "_currentElement": <TopLevelWrapper
                child={
                  <BpkNavigationStack
                    className={null}
                    initialViews={
                      Array [
                        <div />,
                      ]
                    }
                  />
                }
              />,
              "_debugID": 0,
              "_hostContainerInfo": Object {
                "createNodeMock": [Function],
              },
              "_hostParent": null,
              "_instance": TopLevelWrapper {
                "_reactInternalInstance": [Circular],
                "context": Object {},
                "props": Object {
                  "child": <BpkNavigationStack
                    className={null}
                    initialViews={
                      Array [
                        <div />,
                      ]
                    }
                  />,
                },
                "refs": Object {},
                "state": null,
                "updater": Object {
                  "enqueueCallback": [Function],
                  "enqueueCallbackInternal": [Function],
                  "enqueueElementInternal": [Function],
                  "enqueueForceUpdate": [Function],
                  "enqueueReplaceState": [Function],
                  "enqueueSetState": [Function],
                  "isMounted": [Function],
                  "validateCallback": [Function],
                },
              },
              "_mountImage": null,
              "_mountIndex": 0,
              "_mountOrder": 1,
              "_pendingCallbacks": null,
              "_pendingElement": null,
              "_pendingForceUpdate": false,
              "_pendingReplaceState": false,
              "_pendingStateQueue": null,
              "_renderedComponent": [Circular],
              "_renderedNodeType": 1,
              "_rootNodeID": 0,
              "_topLevelWrapper": null,
              "_updateBatchNumber": null,
              "_warnedAboutRefsInRender": false,
            },
            "_updateBatchNumber": null,
            "_warnedAboutRefsInRender": false,
          },
          "context": Object {},
          "props": Object {
            "className": null,
            "initialViews": Array [
              <div />,
            ],
          },
          "refs": Object {},
          "state": Object {
            "views": Array [
              <div />,
            ],
          },
          "updater": Object {
            "enqueueCallback": [Function],
            "enqueueCallbackInternal": [Function],
            "enqueueElementInternal": [Function],
            "enqueueForceUpdate": [Function],
            "enqueueReplaceState": [Function],
            "enqueueSetState": [Function],
            "isMounted": [Function],
            "validateCallback": [Function],
          },
        }
      }
    />
  </div>
</div>
`;
